File: server/model/backend/s3.js
// https://www.npmjs.com/package/aws-sdk
var AWS = require('aws-sdk');


function decode(path){
    let tmp = path.split('/');
    return {
        bucket: tmp.splice(0, 2)[1] || null,
        path: tmp.join('/')
    }
}

function connect(params){
    let config = {
        apiVersion: '2006-03-01',
        accessKeyId: params.access_key_id,
        secretAccessKey: params.secret_access_key,
        signatureVersion: 'v4',
        s3ForcePathStyle: true,
        //sslEnabled: true
    };
    if(params.endpoint){
        config.endpoint = new AWS.Endpoint(params.endpoint);
    }
    var s3 = new AWS.S3(config);
    return Promise.resolve(s3);
}

module.exports = {
    test: function(params){
        return connect(params)
            .then((s3) => {
                return new Promise((done, err) => {
                    s3.listBuckets(function(error, data) {
                        if(error){ err(error) }
                        else{ done(params) }
                    });
                });
            });
    },
    cat: function(path, params, res){
        path = decode(path);
        return connect(params)
            .then((s3) => {
                return Promise.resolve(s3.getObject({
                    Bucket: path.bucket,
                    Key: path.path
                }).on('httpHeaders', function (statusCode, headers) {
                    res.set('content-type', headers['content-type']);
                    res.set('content-length', headers['content-length']);
                    res.set('last-modified', headers['last-modified']);
                }).createReadStream())
            });
    },
    ls: function(path, params){
        if(/\/$/.test(path) === false) path += '/';
        path = decode(path);
        return connect(params)
            .then((s3) => {
                if(path.bucket === null){
                    return new Promise((done, err) => {
                        s3.listBuckets(function(error, data) {
                            if(error){ err(error) }
                            else{
                                let buckets = data.Buckets.map((bucket) => {
                                    return {
                                        name: bucket.Name,
                                        type: 'bucket',
                                        time: new Date(bucket.CreationDate).getTime(),
                                        can_read: true,
                                        can_delete: true,
                                        can_move: false
                                    }
                                });
                                buckets.push({type: 'metadata', name: './', can_create_file: false, can_create_directory: true});
                                done(buckets)
                            }
                        });
                    });
                }else{
                    return new Promise((done, err) => {
                        s3.listObjects({
                            Bucket: path.bucket,
                            Prefix: path.path,
                            Delimiter: '/'
                        }, function(error, data) {
                            if(error){ err(error) }
                            else{
                                let content = data.Contents
                                    .filter((file) => {
                                        return file.Key === path.path? false : true;
                                    })
                                    .map((file) => {
                                        return {
                                            type: 'file',
                                            size: file.Size,
                                            time: new Date(file.LastModified).getTime(),
                                            name: file.Key.split('/').pop()
                                        }
                                    });
                                let folders = data.CommonPrefixes.map((prefix) => {
                                    return {
                                        type: 'directory',
                                        size: 0,
                                        time: null,
                                        name: prefix.Prefix.split('/').slice(-2)[0]
                                    }
                                });
                                return done([].concat(folders, content));
                            }
                        });
                    });
                }
            });
    },
    write: function(path, stream, params){
        path = decode(path);
        return connect(params)
            .then((s3) => {
                return new Promise((done, err) => {
                    s3.upload({
                        Bucket: path.bucket,
                        Key: path.path,
                        Body: stream,
                        ContentLength: stream.byteCount
                    }, function(error, data) {
                        if(error){ err(error) }
                        else{
                            done('ok');
                        }
                    });
                });
            });
    },
    rm: function(path, params){
        path = decode(path);
        return connect(params)
            .then((s3) => {
                return new Promise((done, err) => {
                    s3.listObjects({
                        Bucket: path.bucket,
                        Prefix: path.path
                    }, function(error, obj){
                        if(error){ err(error); }
                        else{
                            Promise.all(obj.Contents.map((file) => {
                                return deleteObject(s3, path.bucket, file.Key)
                            })).then(function(){
                                if(path.path === ''){
                                    s3.deleteBucket({
                                        Bucket: path.bucket
                                    }, function(error){
                                        if(error){ err(error)}
                                        else{ done('ok'); }
                                    });
                                }else{
                                    done('ok');
                                }
                            })
                        }
                    })
                });                
            });

        function deleteObject(s3, bucket, key){
            return new Promise((done, err) => {
                s3.deleteObject({
                    Bucket: bucket,
                    Key: key
                }, function(error, data) {
                    if(error){ err(error) }
                    else{ done('ok') }
                });                        
            })
        }
    },
    mv: function(from, to, params){
        from = decode(from);
        to = decode(to);
     
        return connect(params)
            .then((s3) => {
                return new Promise((done, err) => {
                    s3.copyObject({
                        Bucket: to.bucket,
                        CopySource: from.bucket+'/'+from.path,
                        Key: to.path
                    }, function(error, data) {
                        if(error){ err(error) }
                        else{
                            s3.deleteObject({
                                Bucket: from.bucket,
                                Key: from.path
                            }, function(error){
                                if(error){ err(error) }
                                else{
                                    done('ok');
                                }
                            })
                        }
                    });
                });
            });
    },
    mkdir: function(path, params){
        if(/\/$/.test(path) === false) path += '/';
        path = decode(path);
        return connect(params)
            .then((s3) => {
                return new Promise((done, err) => {
                    if(path.path === ''){
                        s3.createBucket({
                            Bucket: path.bucket
                        }, function(error, data){
                            if(error){ err(error) }
                            else{ done('ok') }
                        });
                    }else{
                        s3.putObject({
                            Bucket: path.bucket,
                            Key: path.path
                        }, function(error, data) {
                            if(error){ err(error) }
                            else{ done('ok') }
                        });
                    }
                });
            })
    },
    touch: function(path, params){
        path = decode(path);
        return connect(params)
            .then((s3) => {
                return new Promise((done, err) => {
                    s3.putObject({
                        Bucket: path.bucket,
                        Key: path.path,
                        Body: ''
                    }, function(error, data) {
                        if(error){ err(error) }
                        else{ done('ok') }
                    });
                });
            })
    }
}

File: server/model/backend/s3.go
package backend

import (
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/credentials"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/s3"
	"github.com/aws/aws-sdk-go/service/s3/s3manager"
	. "github.com/mickael-kerjean/nuage/server/common"
	"io"
	"os"
	"path/filepath"
	"strings"
)

var S3Cache AppCache

type S3Backend struct {
	client *s3.S3
	config *aws.Config
	params map[string]string
}

func init() {
	S3Cache = NewAppCache(2, 1)
}

func NewS3(params map[string]string, app *App) (IBackend, error) {
	if params["region"] == "" {
		params["region"] = "us-east-2"
	}
	config := &aws.Config{
		Credentials:      credentials.NewStaticCredentials(params["access_key_id"], params["secret_access_key"], ""),
		S3ForcePathStyle: aws.Bool(true),
		Region:           aws.String(params["region"]),
	}
	if params["endpoint"] != "" {
		config.Endpoint = aws.String(params["endpoint"])
	}
	backend := &S3Backend{
		config: config,
		params: params,
		client: s3.New(session.New(config)),
	}
	return backend, nil
}

func (s S3Backend) Info() string {
	return "s3"
}

func (s S3Backend) Meta(path string) *Metadata {
	if path == "/" {
		return &Metadata{
			CanCreateFile: NewBool(false),
			CanRename:     NewBool(false),
			CanMove:       NewBool(false),
			CanUpload:     NewBool(false),
		}
	}
	return nil
}

func (s S3Backend) Ls(path string) ([]os.FileInfo, error) {
	p := s.path(path)
	files := make([]os.FileInfo, 0)

	if p.bucket == "" {
		b, err := s.client.ListBuckets(&s3.ListBucketsInput{})
		if err != nil {
			return nil, err
		}
		for _, bucket := range b.Buckets {
			files = append(files, &File{
				FName:   *bucket.Name,
				FType:   "directory",
				FTime:   bucket.CreationDate.UnixNano() / 1000,
				CanMove: NewBool(false),
			})
		}
		return files, nil
	}

	client := s3.New(s.createSession(p.bucket))
	objs, err := client.ListObjects(&s3.ListObjectsInput{
		Bucket:    aws.String(p.bucket),
		Prefix:    aws.String(p.path),
		Delimiter: aws.String("/"),
	})
	if err != nil {
		return nil, err
	}

	for _, object := range objs.Contents {
		files = append(files, &File{
			FName: filepath.Base(*object.Key),
			FType: "file",
			FTime: object.LastModified.UnixNano() / 1000,
			FSize: *object.Size,
		})
	}
	for _, object := range objs.CommonPrefixes {
		files = append(files, &File{
			FName: filepath.Base(*object.Prefix),
			FType: "directory",
		})
	}
	return files, nil
}

func (s S3Backend) Cat(path string) (io.Reader, error) {
	p := s.path(path)
	client := s3.New(s.createSession(p.bucket))

	input := &s3.GetObjectInput{
		Bucket: aws.String(p.bucket),
		Key:    aws.String(p.path),
	}
	if s.params["encryption_key"] != "" {
		input.SSECustomerAlgorithm = aws.String("AES256")
		input.SSECustomerKey = aws.String(s.params["encryption_key"])
	}
	obj, err := client.GetObject(input)
	if err != nil {
		return nil, err
	}

	return obj.Body, nil
}

func (s S3Backend) Mkdir(path string) error {
	p := s.path(path)
	client := s3.New(s.createSession(p.bucket))

	if p.path == "" {
		_, err := client.CreateBucket(&s3.CreateBucketInput{
			Bucket: aws.String(path),
		})
		return err
	}
	_, err := client.PutObject(&s3.PutObjectInput{
		Bucket: aws.String(p.bucket),
		Key:    aws.String(p.path),
	})
	return err
}

func (s S3Backend) Rm(path string) error {
	p := s.path(path)
	client := s3.New(s.createSession(p.bucket))

	if p.bucket == "" {
		return NewError("Doesn't exist", 404)
	}

	objs, err := client.ListObjects(&s3.ListObjectsInput{
		Bucket:    aws.String(p.bucket),
		Prefix:    aws.String(p.path),
		Delimiter: aws.String("/"),
	})
	if err != nil {
		return err
	}
	for _, obj := range objs.Contents {
		_, err := client.DeleteObject(&s3.DeleteObjectInput{
			Bucket: aws.String(p.bucket),
			Key:    obj.Key,
		})
		if err != nil {
			return err
		}
	}
	for _, pref := range objs.CommonPrefixes {
		s.Rm("/" + p.bucket + "/" + *pref.Prefix)
		_, err := client.DeleteObject(&s3.DeleteObjectInput{
			Bucket: aws.String(p.bucket),
			Key:    pref.Prefix,
		})
		if err != nil {
			return err
		}
	}
	if err != nil {
		return err
	}

	if p.path == "" {
		_, err := client.DeleteBucket(&s3.DeleteBucketInput{
			Bucket: aws.String(p.bucket),
		})
		return err
	}
	_, err = client.DeleteObject(&s3.DeleteObjectInput{
		Bucket: aws.String(p.bucket),
		Key:    aws.String(p.path),
	})
	return err
}

func (s S3Backend) Mv(from string, to string) error {
	f := s.path(from)
	t := s.path(to)
	client := s3.New(s.createSession(f.bucket))

	if f.path == "" {
		return NewError("Can't move this", 403)
	}

	input := &s3.CopyObjectInput{
		Bucket:     aws.String(t.bucket),
		CopySource: aws.String(f.bucket + "/" + f.path),
		Key:        aws.String(t.path),
	}
	if s.params["encryption_key"] != "" {
		input.CopySourceSSECustomerAlgorithm = aws.String("AES256")
		input.CopySourceSSECustomerKey = aws.String(s.params["encryption_key"])
		input.SSECustomerAlgorithm = aws.String("AES256")
		input.SSECustomerKey = aws.String(s.params["encryption_key"])
	}

	_, err := client.CopyObject(input)
	if err != nil {
		return err
	}
	return s.Rm(from)
}

func (s S3Backend) Touch(path string) error {
	p := s.path(path)
	client := s3.New(s.createSession(p.bucket))

	if p.bucket == "" {
		return NewError("Can't do that on S3", 403)
	}

	input := &s3.PutObjectInput{
		Body:          strings.NewReader(""),
		ContentLength: aws.Int64(0),
		Bucket:        aws.String(p.bucket),
		Key:           aws.String(p.path),
	}
	if s.params["encryption_key"] != "" {
		input.SSECustomerAlgorithm = aws.String("AES256")
		input.SSECustomerKey = aws.String(s.params["encryption_key"])
	}
	_, err := client.PutObject(input)
	return err
}

func (s S3Backend) Save(path string, file io.Reader) error {
	p := s.path(path)

	if p.bucket == "" {
		return NewError("Can't do that on S3", 403)
	}
	uploader := s3manager.NewUploader(s.createSession(path))
	input := s3manager.UploadInput{
		Body:   file,
		Bucket: aws.String(p.bucket),
		Key:    aws.String(p.path),
	}
	if s.params["encryption_key"] != "" {
		input.SSECustomerAlgorithm = aws.String("AES256")
		input.SSECustomerKey = aws.String(s.params["encryption_key"])
	}
	_, err := uploader.Upload(&input)
	return err
}

func (s S3Backend) createSession(bucket string) *session.Session {
	params := s.params
	params["bucket"] = bucket
	c := S3Cache.Get(params)
	if c == nil {
		res, err := s.client.GetBucketLocation(&s3.GetBucketLocationInput{
			Bucket: aws.String(bucket),
		})
		if err != nil {
			s.config.Region = aws.String("us-east-1")
		} else {
			if res.LocationConstraint == nil {
				s.config.Region = aws.String("us-east-1")
			} else {
				s.config.Region = res.LocationConstraint
			}
		}
		S3Cache.Set(params, s.config.Region)
	} else {
		s.config.Region = c.(*string)
	}

	sess := session.New(s.config)
	return sess
}

type S3Path struct {
	bucket string
	path   string
}

func (s S3Backend) path(p string) S3Path {
	sp := strings.Split(p, "/")
	bucket := ""
	if len(sp) > 1 {
		bucket = sp[1]
	}
	path := ""
	if len(sp) > 2 {
		path = strings.Join(sp[2:], "/")
	}

	return S3Path{
		bucket,
		path,
	}
}
